clippy_lints/src/unitstruct_default_construction.rs

```diff
@@ -0,0 +1,70 @@
+use clippy_utils::{diagnostics::span_lint_and_sugg, is_from_proc_macro, match_def_path, paths};
+use hir::{def::Res, ExprKind};
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Check for construction on unit struct using `default`.
+    ///
+    /// ### Why is this bad?
+    /// This adds code complexity and an unnessesary function call.
+    ///
+    /// ### Example
+    /// ```rust
+    /// #[derive(Default)]
+    /// struct S<T> {
+    ///     _marker: PhantomData<T>
+    /// }
+    ///
+    /// let _: S<i32> = S {
+    ///     _marker: PhantomData::default()
+    /// };
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let _: S<i32> = Something {
+    ///     _marker: PhantomData
+    /// }
+    /// ```
+    #[clippy::version = "1.71.0"]
+    pub UNITSTRUCT_DEFAULT_CONSTRUCTION,
+    complexity,
+    "unit structs can be contructed without calling `default`"
+}
+declare_lint_pass!(UnitstructDefaultConstruction => [UNITSTRUCT_DEFAULT_CONSTRUCTION]);
+
+impl LateLintPass<'_> for UnitstructDefaultConstruction {
+    fn check_expr<'tcx>(&mut self, cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'tcx>) {
+        // make sure we have a call to `Default::default`
+        let hir::ExprKind::Call(fn_expr, &[]) = expr.kind else {return};
```

Comment:
nit: I would use `if_chain` for this row of if/if let + return

Reply:
done